### PR TITLE
Document and clean datatypes rewriter

### PR DESCRIPTION
**Pull request by ajreynol**
Moves code from h to cpp.

The code in the cpp should not be reviewed in this PR (see #1438).

### CODE REVIEW

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 1]
```diff
@@ -0,0 +1,921 @@
...
+    if (index == depth)
+    {
+      return orig;
+    }
+  }
+  else if (n.getNumChildren() > 0)
+  {
+    std::vector<Node> children;
+    bool childChanged = false;
+    for (unsigned i = 0; i < n.getNumChildren(); i++)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 2]
```diff
@@ -0,0 +1,921 @@
...
+#include "theory/datatypes/datatypes_rewriter.h"
+
+namespace CVC4 {
+namespace theory {
+namespace datatypes {
+
+RewriteResponse DatatypesRewriter::postRewrite(TNode in)
+{
+  Trace("datatypes-rewrite-debug") << "post-rewriting " << in << std::endl;
+  if (in.getKind() == kind::APPLY_CONSTRUCTOR)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Use
`Kind k = in.getKind();`
and then
`if (k == ...)` or `switch(k)`

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 3]
```diff
@@ -0,0 +1,921 @@
...
+  else if (in.getKind() == kind::APPLY_TESTER)
+  {
+    return rewriteTester(in);
+  }
+  else if (in.getKind() == kind::DT_SIZE)
+  {
+    if (in[0].getKind() == kind::APPLY_CONSTRUCTOR)
+    {
+      std::vector<Node> children;
+      for (unsigned i = 0; i < in[0].getNumChildren(); i++)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Do not call getNumChildren() in each loop iteration.

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 4]
```diff
@@ -0,0 +1,921 @@
...
+      {
+        if (in[0][i].getType().isDatatype())
+        {
+          children.push_back(
+              NodeManager::currentNM()->mkNode(kind::DT_SIZE, in[0][i]));
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Use `NodeManager nm = ...` `nm->...`

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 5]
```diff
@@ -0,0 +1,921 @@
...
+  }
+  else if (in.getKind() == kind::DT_HEIGHT_BOUND)
+  {
+    if (in[0].getKind() == kind::APPLY_CONSTRUCTOR)
+    {
+      std::vector<Node> children;
+      Node res;
+      Rational r = in[1].getConst<Rational>();
+      Rational rmo = Rational(r - Rational(1));
+      for (unsigned i = 0; i < in[0].getNumChildren(); i++)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Ditto

**Comment by ajreynol** [comment]
I've opened #1438 to track that we should clean/review the moved code from this commit.

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 6]
```diff
@@ -0,0 +1,921 @@
...
+  {
+    if (n1.getOperator() != n2.getOperator())
+    {
+      Trace("datatypes-rewrite-debug") << "Clash operators : " << n1 << " "
+                                       << n2 << " " << n1.getOperator() << " "
+                                       << n2.getOperator() << std::endl;
+      return true;
+    }
+    Assert(n1.getNumChildren() == n2.getNumChildren());
+    for (unsigned i = 0; i < n1.getNumChildren(); i++)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 7]
```diff
@@ -0,0 +1,921 @@
...
+/** get instantiate cons */
+Node DatatypesRewriter::getInstCons(Node n, const Datatype& dt, int index)
+{
+  Assert(index >= 0 && index < (int)dt.getNumConstructors());
+  std::vector<Node> children;
+  children.push_back(Node::fromExpr(dt[index].getConstructor()));
+  Type t = n.getType().toType();
+  for (unsigned i = 0; i < dt[index].getNumArgs(); i++)
+  {
+    Node nc = NodeManager::currentNM()->mkNode(
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Do not call currentNm() in every iteration.

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 8]
```diff
@@ -0,0 +1,921 @@
...
+}
+
+int DatatypesRewriter::isInstCons(Node t, Node n, const Datatype& dt)
+{
+  if (n.getKind() == kind::APPLY_CONSTRUCTOR)
+  {
+    int index = Datatype::indexOf(n.getOperator().toExpr());
+    const DatatypeConstructor& c = dt[index];
+    Type nt = n.getType().toType();
+    for (unsigned i = 0; i < n.getNumChildren(); i++)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 9]
```diff
@@ -0,0 +1,921 @@
...
+#else
+    return n.eqNode(DatatypesRewriter::getInstCons(n, dt, i));
+#endif
+  }
+}
+
+bool DatatypesRewriter::isNullaryApplyConstructor(Node n)
+{
+  Assert(n.getKind() == kind::APPLY_CONSTRUCTOR);
+  for (unsigned i = 0; i < n.getNumChildren(); i++)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 10]
```diff
@@ -0,0 +1,921 @@
...
+    {
+      return false;
+    }
+  }
+  return true;
+}
+
+bool DatatypesRewriter::isNullaryConstructor(const DatatypeConstructor& c)
+{
+  for (unsigned j = 0; j < c.getNumArgs(); j++)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 11]
```diff
@@ -0,0 +1,921 @@
...
+    }
+    // now run DFA minimization on term structure
+    Trace("dt-nconst") << "  " << terms.size()
+                       << " total subterms :" << std::endl;
+    int eqc_count = 0;
+    std::map<Node, int> eqc_op_map;
+    std::map<Node, int> eqc;
+    std::map<int, std::vector<Node> > eqc_nodes;
+    // partition based on top symbol
+    for (unsigned j = 0; j < terms.size(); j++)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 12]
```diff
@@ -0,0 +1,921 @@
...
+        {
+          // look at all nodes in this equivalence class
+          unsigned nchildren = eqc_nodes[eqc_curr][0].getNumChildren();
+          std::map<int, std::vector<Node> > prt;
+          for (unsigned j = 0; j < nchildren; j++)
+          {
+            prt.clear();
+            // partition based on children : for the first child that causes a
+            // split, break
+            for (unsigned k = 0; k < eqc_nodes[eqc_curr].size(); k++)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 13]
```diff
@@ -0,0 +1,921 @@
...
+              break;
+            }
+          }
+          // move into new eqc(s)
+          for (std::map<int, std::vector<Node> >::iterator it = prt.begin();
+               it != prt.end();
+               ++it)
+          {
+            int e = eqc_count;
+            for (unsigned j = 0; j < it->second.size(); j++)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 14]
```diff
@@ -0,0 +1,921 @@
...
+  {
+    if (tn.isCodatatype())
+    {
+      return normalizeCodatatypeConstant(n);
+    }
+    else
+    {
+      std::vector<Node> children;
+      bool childrenChanged = false;
+      for (unsigned i = 0; i < n.getNumChildren(); i++)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 15]
```diff
@@ -0,0 +1,921 @@
...
+    for (std::map<Node, Node>::iterator it = rf.begin(); it != rf.end(); ++it)
+    {
+      Trace("dt-nconst-debug") << "Mapping equivalence class of " << it->first
+                               << " -> " << it->second << std::endl;
+      Assert(eqc.find(it->second) != eqc.end());
+      eqc[it->first] = eqc[it->second];
+    }
+    // we now have a partition of equivalent terms
+    Trace("dt-nconst") << "Computed equivalence classes ids : " << std::endl;
+    for (std::map<Node, int>::iterator it = eqc.begin(); it != eqc.end(); ++it)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Range-based loop?

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 16]
```diff
@@ -0,0 +1,921 @@
...
+    {
+      isCdt = true;
+      if (n.getKind() == kind::APPLY_CONSTRUCTOR)
+      {
+        sk.push_back(n);
+        rf_pending.push_back(Node::null());
+        std::vector<Node> children;
+        children.push_back(n.getOperator());
+        bool childChanged = false;
+        for (unsigned i = 0; i < n.getNumChildren(); i++)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 17]
```diff
@@ -0,0 +1,921 @@
...
+      int debruijn = depth - it->second - 1;
+      return NodeManager::currentNM()->mkConst(
+          UninterpretedConstant(n.getType().toType(), debruijn));
+    }
+    else
+    {
+      std::vector<Node> children;
+      bool childChanged = false;
+      eqc_stack[e] = depth;
+      for (unsigned i = 0; i < n.getNumChildren(); i++)
```
Comment (aniemetz) [inline src/theory/datatypes/datatypes_rewriter.cpp]: Ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 18]
```diff
@@ -0,0 +1,921 @@
...
+      for (unsigned i = 0; i < in[0].getNumChildren(); i++)
+      {
+        if (in[0][i].getType().isDatatype())
+        {
+          if (r.isZero())
+          {
+            res = NodeManager::currentNM()->mkConst(false);
+            break;
+          }
+          else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: spurious ``else``

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 19]
```diff
@@ -0,0 +1,921 @@
...
+  }
+
+  if (in.getKind() == kind::EQUAL)
+  {
+    if (in[0] == in[1])
+    {
+      return RewriteResponse(REWRITE_DONE,
+                             NodeManager::currentNM()->mkConst(true));
+    }
+    else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 20]
```diff
@@ -0,0 +1,921 @@
...
+        //  return RewriteResponse(REWRITE_AGAIN_FULL, rew[0] );
+      }
+      else if (in[1] < in[0])
+      {
+        Node ins = NodeManager::currentNM()->mkNode(in.getKind(), in[1], in[0]);
+        Trace("datatypes-rewrite") << "Swap equality " << in << " to " << ins
+                                   << std::endl;
+        return RewriteResponse(REWRITE_DONE, ins);
+      }
+      else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 21]
```diff
@@ -0,0 +1,921 @@
...
+    // constant may be a subterm of another constant, so cannot assume that this
+    // will succeed for codatatypes
+    // Assert( !inn.isNull() );
+    if (!inn.isNull() && inn != in)
+    {
+      Trace("datatypes-rewrite") << "Normalized constant " << in << " -> "
+                                 << inn << std::endl;
+      return RewriteResponse(REWRITE_DONE, inn);
+    }
+    else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 22]
```diff
@@ -0,0 +1,921 @@
...
+  {
+    bool result = Datatype::indexOf(in.getOperator().toExpr())
+                  == Datatype::indexOf(in[0].getOperator().toExpr());
+    Trace("datatypes-rewrite") << "DatatypesRewriter::postRewrite: "
+                               << "Rewrite trivial tester " << in << " "
+                               << result << std::endl;
+    return RewriteResponse(REWRITE_DONE,
+                           NodeManager::currentNM()->mkConst(result));
+  }
+  else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 23]
```diff
@@ -0,0 +1,921 @@
...
+  }
+  else if (n1 != n2)
+  {
+    if (n1.isConst() && n2.isConst())
+    {
+      Trace("datatypes-rewrite-debug") << "Clash constants : " << n1 << " "
+                                       << n2 << std::endl;
+      return true;
+    }
+    else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 24]
```diff
@@ -0,0 +1,921 @@
...
+  return false;
+}
+/** get instantiate cons */
+Node DatatypesRewriter::getInstCons(Node n, const Datatype& dt, int index)
+{
+  Assert(index >= 0 && index < (int)dt.getNumConstructors());
+  std::vector<Node> children;
+  children.push_back(Node::fromExpr(dt[index].getConstructor()));
+  Type t = n.getType().toType();
+  for (unsigned i = 0; i < dt[index].getNumArgs(); i++)
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: cache size

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 25]
```diff
@@ -0,0 +1,921 @@
...
+}
+
+Node DatatypesRewriter::mkTester(Node n, int i, const Datatype& dt)
+{
+  if (options::dtUseTesters())
+  {
+    return NodeManager::currentNM()->mkNode(
+        kind::APPLY_TESTER, Node::fromExpr(dt[i].getTester()), n);
+  }
+  else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 26]
```diff
@@ -0,0 +1,921 @@
...
+    for (std::map<Node, int>::iterator it = eqc.begin(); it != eqc.end(); ++it)
+    {
+      Trace("dt-nconst") << "  " << it->first << " -> " << it->second
+                         << std::endl;
+    }
+    // traverse top-down based on equivalence class
+    std::map<int, int> eqc_stack;
+    return normalizeCodatatypeConstantEqc(s, eqc_stack, eqc, 0);
+  }
+  else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 27]
```diff
@@ -0,0 +1,921 @@
...
+      {
+        Node nc = normalizeConstant(n[i]);
+        children.push_back(nc);
+        childrenChanged = childrenChanged || nc != n[i];
+      }
+      if (childrenChanged)
+      {
+        return NodeManager::currentNM()->mkNode(n.getKind(), children);
+      }
+      else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 28]
```diff
@@ -0,0 +1,921 @@
...
+      {
+        return n;
+      }
+    }
+  }
+  else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 29]
```diff
@@ -0,0 +1,921 @@
...
+          {
+            return Node::null();
+          }
+          childChanged = nc != n[i] || childChanged;
+          children.push_back(nc);
+        }
+        sk.pop_back();
+        if (childChanged)
+        {
+          ret = NodeManager::currentNM()->mkNode(kind::APPLY_CONSTRUCTOR,
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: cache NM outside for

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 30]
```diff
@@ -0,0 +1,921 @@
...
+      else
+      {
+        // a loop
+        const Integer& i = n.getConst<UninterpretedConstant>().getIndex();
+        uint32_t index = i.toUnsignedInt();
+        if (index >= sk.size())
+        {
+          return Node::null();
+        }
+        else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 31]
```diff
@@ -0,0 +1,921 @@
...
+        {
+          Assert(sk.size() == rf_pending.size());
+          Node r = rf_pending[rf_pending.size() - 1 - index];
+          if (r.isNull())
+          {
+            r = NodeManager::currentNM()->mkBoundVar(
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 32]
```diff
@@ -0,0 +1,921 @@
...
+Node DatatypesRewriter::normalizeCodatatypeConstantEqc(
+    Node n, std::map<int, int>& eqc_stack, std::map<Node, int>& eqc, int depth)
+{
+  Trace("dt-nconst-debug") << "normalizeCodatatypeConstantEqc: " << n
+                           << " depth=" << depth << std::endl;
+  if (eqc.find(n) == eqc.end())
+  {
+    return n;
+  }
+  else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 33]
```diff
@@ -0,0 +1,921 @@
...
+  {
+    int e = eqc[n];
+    std::map<int, int>::iterator it = eqc_stack.find(e);
+    if (it != eqc_stack.end())
+    {
+      int debruijn = depth - it->second - 1;
+      return NodeManager::currentNM()->mkConst(
+          UninterpretedConstant(n.getType().toType(), debruijn));
+    }
+    else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

File: src/theory/datatypes/datatypes_rewriter.cpp [hunk 34]
```diff
@@ -0,0 +1,921 @@
...
+        childChanged = childChanged || nc != n[i];
+      }
+      eqc_stack.erase(e);
+      if (childChanged)
+      {
+        Assert(n.getKind() == kind::APPLY_CONSTRUCTOR);
+        children.insert(children.begin(), n.getOperator());
+        return NodeManager::currentNM()->mkNode(n.getKind(), children);
+      }
+      else
```
Comment (HanielB) [inline src/theory/datatypes/datatypes_rewriter.cpp]: ditto

**Pull request merged**